### ISSUE TITLE
Discussion - Names of embedded schema GQL Types

### ISSUE THREAD

**Issue by yoadsn** [nodkz/graphql-compose-mongoose#23]
Assume I have this schema:

```
let imageDataStructure = Schema({
  url: String,
  creditsTo: String,
  creditsLink: String,
  dimensions : {
    aspectRatio: Number
  },
  metadata: {
    leadColor: String,
    gravityCenter: {
      x: Number,
      y: Number
    }
  }
}, { _id: false });

let homeSchema = Schema({
  address: String,
  streetImage: imageDataStructure
})

let agentSchema = Schema({
  first: String,
  last: String,
  profileImage: imageDataStructure
})
```

The generated GQL types (depending on order of schema parsing) would be:
```
homeSchema.streetImage -> HomeStreetImage
agentSchema.profileImage -> HomeStreetImage
```
The type is reused since it's the same mongoose schema. The name is picked by the first path that discovers this schema. All of that takes place in `fieldsConverter.js` around `embeddedToGraphQL`.

Now if I would create the image structure like so:
```
let imageDataStructure = {
  url: String,
  creditsTo: String,
  creditsLink: String,
  dimensions : {
    aspectRatio: Number
  },
  metadata: {
    leadColor: String,
    gravityCenter: {
      x: Number,
      y: Number
    }
  }
};

... reset is the same
```

the following GQL types are expected:
```
homeSchema.streetImage -> HomeStreetImage
agentSchema.profileImage -> AgentProfileImage
```
It's not the same mongoose schema - and so a new Type would be generated for each path.

Since the structure of the embedded image data is the same. I would assume it's useful to create GQL fragments for this type and use it where ever a type contains this embedded type. Something like this:
```
query {
  bestAgent {
    first last
    profileImage {
      ... basicImageData
    }
  }
  bestHome {
    address
    streetImage {
      ... basicImageData
    }
  }
}
fragment basicImageData on ImageData {

  url
  dimensions {
    aspectRatio
  }
}
```

My idea is that since when using an embedded Schema (not a plain JS object) for reuse in mongoose the name of the type is anyway going to be wrong for any other types except for the first path that discovered this type - We might want to document the best way to force the name in advance.

Calling `composeWithMongoose` on the image model is not solving - I guess because it's not the same schema once you create a Model from it?

Calling something like this before all other `composeWithMongoose`:
```
convertSchemaToGraphQL(imageDataStructure, 'EmbeddedImage');
```
Would do the trick by placing a TC on the schema object `_gqcTypeComposer` prop - but does it feel like a hack?

Love to hear your thought.

**Comment by nodkz** [comment]
It's not a hack. I'm using this in the same way. When I want to provide force name for embedded schema I'm using `convertSchemaToGraphQL`.
 
```js
import { Schema } from 'mongoose';
import { convertSchemaToGraphQL } from 'graphql-compose-mongoose';

export const LonLatSchema = new Schema(
  {
    lng: {
      type: Number,
    },
    lat: {
      type: Number,
    },
  },
  {
    _id: false,
  }
);

convertSchemaToGraphQL(LonLatSchema, 'LonLat');
```

**Comment by yoadsn** [comment]
@nodkz Thanks, Good to know.
I'm closing this issue let me know if you think I should add anything about this to the docs.

**Comment by nodkz** [comment]
@yoadsn yep. It will be cool.

Also I'm very impressed that you have such deep look on internals of graphql-compose-...
👍

PS. I finished [elasticsearch](https://github.com/nodkz/graphql-compose-elasticsearch) before my vacation and it in pre-production in our internal app. So may take a look on it. So many crazy things inside. `graphql-compose` is used at full power.

**Comment by janfabian** [comment]
Hi, thanks for sharing this thought. If I want to create multiple TC from one model, is this safe?
```
let OccupationTC = composeWithMongoose(OccupationModel);
OccupationModel.schema._gqcTypeComposer = null;
let occupationStaffTC = composeWithMongoose(OccupationModel, {
  name: "OccupationStaff"
});
```

**Comment by janfabian** [comment]
There is trouble with embedded documents, which was in fact the original question. 

```javascript
let BillSchema = {
  amount: Number
};
let occupationScheme = {
  bill: BillSchema
};
```

if I then try to generate TC in the following way 

```javascript
let OccupationTC = composeWithMongoose(OccupationModel);
OccupationModel.schema._gqcTypeComposer = null;
let occupationStaffTC = composeWithMongoose(OccupationModel, {
  name: "OccupationStaff"
});
```

It will fail, because `Error: Schema must contain unique named types but contains multiple types named "Bill".`

The solution is to remove embedded fields and add them manually later
```javascript
let OccupationTC = composeWithMongoose(OccupationModel);
OccupationModel.schema._gqcTypeComposer = null;
let occupationStaffTC = composeWithMongoose(OccupationModel, {
  name: "OccupationStaff",
  fields: {
    remove: ["bill"]
  }
});
occupationStaffTC.addFields({
  bill: convertModelToGraphQL(BillModel, "BillStaff"),
});
```

**Comment by yoadsn** [comment]
@janfabian I am not sure about your usage of nullifying an internal library field (yes, you can read the source and know about it, but no, it is not part of the library API).
Though I totally understand you want to work around a problem here, let's try and improve the API of library if possible..

Can you give a bit more information about the use-case? Why would you like to create two different TC's from the same Model?

- Do they share the structure of two different entities? (in that case I think you need two models, not two TCs for the same Model)
- Are you going to customize the two TC's in different ways? In that case we need to think about cloning a TC rather then generating two.

As you can see, the relation between a Model (actually the Schema of the model) and a TC is 1-1 by the way the library stores the reference to the TC on the internal field. If this approach is broken with your use-case I would really like to know it better.

As to the error you get - this is generated by GraphQL and the "schema" mentioned is not the mongoose schema but the entire GQL schema. It does not allow to redefine the same type more than once. (and rightfully so).

Why does that happen?
My assumption:

Upon your first call
```
let OccupationTC = composeWithMongoose(OccupationModel);
```
the type storage of GQL compose would parse the type tree, create a TC for the embedded "Bill" schema and place it on the `_gqcTypeComposer` field, then create a TC for the "Occupation" schema and store it on the `_gqcTypeComposer` field.
Then, you nullify `_gqcTypeComposer` on the "Occupation" and call it again with a different name:
```
let occupationStaffTC = composeWithMongoose(OccupationModel, {
  name: "OccupationStaff"
});
```
Which allows GQC to convert the same tree again. When it arrives at the "Bill" schema it will reuse the TC already created (with the type name already set from the previous tree) and not set the correct "prefix" on the type name (prefix is ignored in fact since the TC is reused).

This would lead to multiple types sent to GraphQL with the same name

Prefix was your guard against this error - making the ability to reuse the same Schema in different parent types and getting different type names. Messing up `_gqcTypeComposer` too this guard away.

P.S since the error give the name "Bill" - I must assume you also called `composeWithMongoose` on a "Bill" Model before this code and this is not the complete example - otherwise I'm getting something wrong here.

**Comment by janfabian** [comment]
@yoadsn Thank you for this detailed info! It helps me to summarize what I'm trying to achieve. 

That is following: I have my entity Occupation. Then I have a set of authorisation roles - customer, staff, manager etc. Each role should be able to see this Occupation entity, but I need to restrict fields (or args, relations) which are displayed to each role. Customer should not see some internal information about the occupation which is only for staff. 

Therefore I want to have OccupationCustomerTC, OccupationStaffTC etc. to manage and use them separately. But the model in db is still one - Occupation. 

The answer for this might be to use `clone` as you suggest. But I have following issue with this approach. 

```javascript
let OccupationTC = composeWithMongoose(OccupationModel);
let occupationStaffTC = OccupationTC.clone("OccupationStaff");
```

And somewhere else I use

```javascript
SomeOtherTC.addRelation(
  "occupations",
  () => ({
    resolver: occupationStaff.get('$findMany')
{resolver: occupationStaff.get('$findMany')}
```

**Comment by yoadsn** [comment]
@janfabian I can never hope to fully understand your business use-case over an issue thread but I will try my best.

To me, it seems the problem of "data store and access" and the problem of "authorization" should be solved separately. The **Occupation** entity is the same regardless of which staff accesses it. I would argue **Occupation** may grow (more fields, more relations) in the future regardless of authorization and the other way around, authorization may change without the **Occupation** entity changing.

This hints that the implementation should also consider how tying the data access model of **Occupation** to the authorization model could complicate your implementation going forward.

My approach would be to have a single **Occupation** data model and data store representation. The authorization rules should be implemented as a "filter" on the resolvers. Depending on how complex / generic you want to make it it could have this logic:
- Context knows which auth roles are active and merges them
- Each resolver of the **Occupation** returns a redacted version based on the auth rules. This could either be changing the return value from the DB (less secure) or changing the query itself (more secure).
- Client of the API would have knowledge of the Auth and would "filter the queries" in the same manner to not even request the fields which are not allowed.
This is mostly to improve the user/developer experience - the actual line of defense is the server filtering.

Of course - you must decide first if your auth model is a blacklist (tell what's not allowed) or a whitelist (tell only what's allowed). Each has it's own pros and cons.

Around this basic approach you can build stronger concepts - like Auth-to-AST mapping schema that could generate the filters from Auth metadata. Using MongoDB's aggregation pipeline stage "redact" for applying the Auth on the db side for larger reports and so on..

Bottom line - I feel Data Model should be orthogonal to the  auth model. Would love more experienced guys reading this to comment.

**Comment by janfabian** [comment]
@yoadsn 
> Context knows which auth roles are active and merges them

I agree, this is solved in the context of web framework (Koa in my case). Then in the first layer of resolve 'onion' I check the authorization role. This is OK. The user is not allowed to see anything, if the needed role is not assigned to him.

> This could either be changing the return value from the DB (less secure) or changing the query itself (more secure)

Do you propose to create specific resolvers for this ?

```javascript
let occupationStaffFindById = OccupationTC.
  get("$findById").
  wrap((newResolver) => {
    // ... newResolver.removeFields
    // ... newResolver.extendField etc?
  });
```

or 

```javascript
let occupationStaffFindById = OccupationTC.
  get("$findById").
  wrapResolve((next) => (rp) => {
    // ... edit rp.projection
  });
```

**Comment by yoadsn** [comment]
@janfabian More like the second option.
There are cases where you would lieke the GraphQL schema to be different (for example different type, different field/args) for different users. But that would be the edge case in my opinion.
Same structure for all, data is redacted as needed.

**Comment by janfabian** [comment]
thanks @yoadsn for your comments and insights
